libsql-server/src/connection/libsql.rs

```diff
@@ -621,61 +642,71 @@ impl<W: Wal> Connection<W> {
-                        .slot
-                        .take()
-                        .expect("there should be a slot right after downgrading a txn");
-                    let mut maybe_state_slot = state.slot.write();
-                    // We need to make sure that the state slot is our slot before removing it.
-                    if let Some(ref state_slot) = *maybe_state_slot {
-                        if Arc::ptr_eq(state_slot, &old_slot) {
-                            maybe_state_slot.take();
-                        }
-                    }
-
-                    drop(maybe_state_slot);
-
-                    state.notify.notify_waiters();
-                }
-                // nothing to do
-                (_, _) => (),
-            }
-
-            previous_state = new_state;
+            let ret = lock.execute_step(step, &results, &mut builder);
+            // /!\ always make sure that the state is updated before returning
+            previous_state = lock.update_state(this.clone(), previous_state, txn_timeout)?;
+            let res = ret?;
 
             results.push(res);
         }
 
         {
             let mut lock = this.lock();
             let is_autocommit = lock.conn.is_autocommit();
-            builder.finish(
-                *(lock.current_frame_no_receiver.borrow_and_update()),
-                is_autocommit,
-            )?;
+            let current_fno = *lock.current_frame_no_receiver.borrow_and_update();
+            builder.finish(current_fno, is_autocommit)?;
         }
 
         Ok(builder)
     }
 
+    fn update_state(
+        &mut self,
+        arc_this: Arc<Mutex<Self>>,
+        previous_state: TransactionState,
+        txn_timeout: Duration,
+    ) -> Result<TransactionState> {
+        use rusqlite::TransactionState as Tx;
+
+        let new_state = self.conn.transaction_state(Some(DatabaseName::Main))?;
+        match (previous_state, new_state) {
+            // lock was upgraded, claim the slot
+            (Tx::None | Tx::Read, Tx::Write) => {
+                let slot = Arc::new(TxnSlot {
+                    conn: arc_this,
+                    created_at: Instant::now(),
+                    is_stolen: false.into(),
+                    txn_timeout,
+                });
+
+                self.slot.replace(slot.clone());
+                self.state.slot.write().replace(slot);
+            }
+            // lock was downgraded, notify a waiter
+            (Tx::Write, Tx::None | Tx::Read) => {
+                let old_slot = self
+                    .slot
+                    .take()
+                    .expect("there should be a slot right after downgrading a txn");
+                let mut maybe_state_slot = self.state.slot.write();
+                // We need to make sure that the state slot is our slot before removing it.
+                if let Some(ref state_slot) = *maybe_state_slot {
+                    if Arc::ptr_eq(state_slot, &old_slot) {
```

Comment:
is there a case where a pointer gets reused and we run into a ABA problem?

Reply:
no, we discussed that last time, but we're actually comparing the pointer with itself, and we're still holding a reference to both ends while comparing, so no risk of it being reused in the meantime :)